proxy_multi_map.go

```diff
@@ -0,0 +1,515 @@
+		switch event.EventType {
+		case hazelcast.EntryAdded:
+			fmt.Println("Entry Added:", event.Value)
+		case hazelcast.EntryRemoved:
+			fmt.Println("Entry Removed:", event.Value)
+		case hazelcast.EntryUpdated:
+			fmt.Println("Entry Updated:", event.Value)
+		case hazelcast.EntryEvicted:
+			fmt.Println("Entry Remove:", event.Value)
+		case hazelcast.EntryLoaded:
+			fmt.Println("Entry Loaded:", event.Value)
+		}
+	})
+
+Adding an event listener returns a subscription ID, which you can later use to remove the listener:
+
+	err = m.RemoveEntryListener(ctx, subscriptionID)
+
+Using Locks
+
+You can lock entries in a MultiMap.
+When an entry is locked, only the owner of that lock can access that entry in the cluster until it is unlocked by the owner of force unlocked.
+See https://docs.hazelcast.com/imdg/latest/data-structures/map.html#locking-maps for details, usage is identical.
+
+Locks are reentrant.
+The owner of a lock can acquire the lock again without waiting for the lock to be unlocked.
+If the key is locked N times, it should be unlocked N times before another goroutine can acquire it.
+
+Lock ownership in Hazelcast Go Client is explicit.
+The first step to own a lock is creating a lock context, which is similar to a key.
+The lock context is a regular context.Context which carry a special value that uniquely identifies the lock context in the cluster.
+Once the lock context is created, it can be used to lock/unlock entries and used with any function that is lock aware, such as Put.
+
+	m, err := client.GetMultiMap(ctx, "my-map")
+	lockCtx := m.NewLockContext(ctx)
+	// block acquiring the lock
+	err = m.Lock(lockCtx, "some-key")
+	// pass lock context to use the locked entry
+	err = m.Put(lockCtx, "some-key", "some-value")
+	// release the lock once done with it
+	err = m.Unlock(lockCtx, "some-key")
+
+As mentioned before, lock context is a regular context.Context which carry a special lock ID.
+You can pass any context.Context to any MultiMap function, but in that case lock ownership between operations using the same hazelcast.Client instance is not possible.
+
+*/
+type MultiMap struct {
+	*proxy
+}
+
+func newMultiMap(p *proxy) *MultiMap {
+	return &MultiMap{proxy: p}
+}
+
+// NewLockContext augments the passed parent context with a unique lock ID.
+// If passed context is nil, context.Background is used as the parent context.
+func (m *MultiMap) NewLockContext(ctx context.Context) context.Context {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	// todo this is shared with map, should we use a different lockIDKey?
+	return context.WithValue(ctx, lockIDKey, lockID(m.refIDGen.NextID()))
+}
+
+// AddEntryListener adds a continuous entry listener to this map.
+func (m *MultiMap) AddEntryListener(ctx context.Context, config MultiMapEntryListenerConfig, handler EntryNotifiedHandler) (types.UUID, error) {
+	//todo does smart field corresponds to "localOnly"?
```

Comment:
That's right. In smart mode, the member does not send the request to another member (so it is local). In unisocket mode, it forwards the request to other members.